File: dotfiles-ai/skills/openrouter_free_models/openrouter_free_models.ts
/**
 * OpenRouter Free Models Manager (TypeScript)
 *
 * Fetches all models from OpenRouter, filters free models,
 * and tests each one to maintain an active list of working free models.
 *
 * Usage:
 *   import { OpenRouterFreeModels } from './openrouter_free_models';
 *
 *   const manager = new OpenRouterFreeModels();
 *   await manager.refreshAllModels();
 *   await manager.testFreeModels();
 *   const active = manager.getActiveFreeModels();
 */

import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

export interface ModelInfo {
  id: string;
  name: string;
  context_length: number;
  pricing_prompt: number;
  pricing_completion: number;
  architecture: Record<string, unknown>;
  top_provider: Record<string, unknown>;
  supported_parameters: string[];
  created: number;
  description?: string;
  hugging_face_id?: string | null;
}

export interface ActiveFreeModel {
  model_id: string;
  name: string;
  context_length: number;
  tested_at: number;
  response_time_ms: number;
  test_prompt: string;
  test_response: string;
  status: 'active' | 'failed';
}

export interface FailedModel {
  model_id: string;
  name: string;
  context_length: number;
  http_code?: string;
  error?: string;
}

export interface TestResults {
  active: ActiveFreeModel[];
  failed: FailedModel[];
  total_tested: number;
  active_count: number;
  failed_count: number;
}

export interface ModelSummary {
  total_models: number;
  free_models: number;
  active_free_models: number;
  last_refresh: string;
  last_test: string;
}

export class OpenRouterFreeModels {
  private static readonly BASE_URL = 'https://openrouter.ai/api/v1';
  private static readonly MODELS_ENDPOINT = `${OpenRouterFreeModels.BASE_URL}/models`;
  private static readonly CHAT_ENDPOINT = `${OpenRouterFreeModels.BASE_URL}/chat/completions`;
  private static readonly TEST_PROMPT = 'Reply with exactly: PING';
  private static readonly TEST_TIMEOUT = 30000;
  private static readonly CONCURRENT_TESTS = 5;

  private apiKey: string;
  cacheFile: string;
  activeCacheFile: string;
  failedCacheFile: string;

  allModels: ModelInfo[] = [];
  freeModels: ModelInfo[] = [];
  activeFreeModels: ActiveFreeModel[] = [];
  failedModels: FailedModel[] = [];

  private lastRefresh = 0;
  private lastTest = 0;

  constructor(apiKey?: string, cacheDir?: string) {
    this.apiKey = apiKey || process.env.OPENROUTER_API_KEY || '';
    if (!this.apiKey) {
      throw new Error(
        'OPENROUTER_API_KEY not set. Provide via constructor or environment variable.'
      );
    }

    const defaultCacheDir = path.join(os.homedir(), '.cache', 'openrouter');
    const dir = cacheDir || defaultCacheDir;

    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    this.cacheFile = path.join(dir, 'all_models.json');
    this.activeCacheFile = path.join(dir, 'active_free_models.json');
    this.failedCacheFile = path.join(dir, 'failed_free_models.json');
  }

  private getHeaders(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
      'HTTP-Referer': 'https://github.com/cbwinslow/dotfiles',
      'X-Title': 'CBW OpenRouter Free Models',
    };
  }

  private async fetchJson<T>(url: string, options?: RequestInit): Promise<T> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), OpenRouterFreeModels.TEST_TIMEOUT);

    try {
      const response = await fetch(url, {
        ...options,
        signal: controller.signal,
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      return (await response.json()) as T;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async refreshAllModels(): Promise<ModelInfo[]> {
    interface ModelsResponse {
      data: Array<{
        id: string;
        name: string;
        context_length: number;
        pricing: { prompt: string | number; completion: string | number; [key: string]: unknown };
        architecture: Record<string, unknown>;
        top_provider: Record<string, unknown>;
        supported_parameters: string[];
        created: number;
        description?: string;
        hugging_face_id?: string | null;
        [key: string]: unknown;
      }>;
    }

    const data = await this.fetchJson<ModelsResponse>(OpenRouterFreeModels.MODELS_ENDPOINT, {
      headers: this.getHeaders(),
    });

    this.allModels = data.data.map((m) => ({
      id: m.id,
      name: m.name || m.id,
      context_length: m.context_length || 0,
      pricing_prompt: Number(m.pricing?.prompt || 0),
      pricing_completion: Number(m.pricing?.completion || 0),
      architecture: m.architecture || {},
      top_provider: m.top_provider || {},
      supported_parameters: m.supported_parameters || [],
      created: m.created || 0,
      description: m.description || '',
      hugging_face_id: m.hugging_face_id || null,
    }));

    this.filterFreeModels();
    this.saveCache();
    this.lastRefresh = Date.now();

    return this.freeModels;
  }

  private filterFreeModels(): void {
    this.freeModels = this.allModels.filter(
      (m) => m.pricing_prompt === 0 && m.pricing_completion === 0
    );
  }

  private saveCache(): void {
    const cache = {
      refreshed_at: this.lastRefresh,
      total_models: this.allModels.length,
      free_models_count: this.freeModels.length,
      free_models: this.freeModels,
    };
    fs.writeFileSync(this.cacheFile, JSON.stringify(cache, null, 2));
  }

  private saveActiveCache(): void {
    const cache = {
      tested_at: this.lastTest,
      active_count: this.activeFreeModels.length,
      failed_count: this.failedModels.length,
      active_models: this.activeFreeModels,
      failed_models: this.failedModels,
    };
    fs.writeFileSync(this.activeCacheFile, JSON.stringify(cache.active_models, null, 2));
    fs.writeFileSync(this.failedCacheFile, JSON.stringify(cache.failed_models, null, 2));
  }

  private async testSingleModel(
    modelId: string,
    name: string,
    contextLength: number
  ): Promise<ActiveFreeModel | null> {
    const startTime = Date.now();

    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), OpenRouterFreeModels.TEST_TIMEOUT);

      const response = await fetch(OpenRouterFreeModels.CHAT_ENDPOINT, {
        method: 'POST',
        headers: this.getHeaders(),
        signal: controller.signal,
        body: JSON.stringify({
          model: modelId,
          messages: [{ role: 'user', content: OpenRouterFreeModels.TEST_PROMPT }],
          max_tokens: 10,
        }),
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        return null;
      }

      const data = (await response.json()) as {
        choices: Array<{ message: { content: string } }>;
      };

      const elapsedMs = Date.now() - startTime;
      const choice = data.choices?.[0];
      const message = choice?.message as Record<string, unknown> | undefined;
      const content = (message?.content as string) || (message?.reasoning as string) || '';

      return {
        model_id: modelId,
        name,
        context_length: contextLength,
        tested_at: Date.now(),
        response_time_ms: Math.round(elapsedMs * 100) / 100,
        test_prompt: OpenRouterFreeModels.TEST_PROMPT,
        test_response: content.substring(0, 200),
        status: 'active',
      };
    } catch {
      return null;
    }
  }

  async testFreeModels(options?: {
    modelIds?: string[];
    maxConcurrent?: number;
  }): Promise<TestResults> {
    if (this.freeModels.length === 0) {
      await this.refreshAllModels();
    }

    const targets = options?.modelIds
      ? this.freeModels.filter((m) => options.modelIds!.includes(m.id))
      : [...this.freeModels];

    const maxConcurrent = options?.maxConcurrent ?? OpenRouterFreeModels.CONCURRENT_TESTS;

    this.activeFreeModels = [];
    this.failedModels = [];

    // Process with concurrency control
    const results: (ActiveFreeModel | null)[] = [];
    const queue = [...targets];

    const worker = async () => {
      while (queue.length > 0) {
        const model = queue.shift()!;
        const result = await this.testSingleModel(model.id, model.name, model.context_length);
        results.push(result);
      }
    };

    const workers = Array.from({ length: Math.min(maxConcurrent, targets.length) }, () => worker());
    await Promise.all(workers);

    for (let i = 0; i < targets.length; i++) {
      const model = targets[i];
      const result = results[i];

      if (result) {
        this.activeFreeModels.push(result);
      } else {
        this.failedModels.push({
          model_id: model.id,
          name: model.name,
          context_length: model.context_length,
        });
      }
    }

    this.lastTest = Date.now();
    this.saveActiveCache();

    return {
      active: this.activeFreeModels,
      failed: this.failedModels,
      total_tested: targets.length,
      active_count: this.activeFreeModels.length,
      failed_count: this.failedModels.length,
    };
  }

  getActiveFreeModels(): ActiveFreeModel[] {
    return this.activeFreeModels;
  }

  getFreeModelIds(): string[] {
    return this.freeModels.map((m) => m.id);
  }

  getModelById(modelId: string): ModelInfo | undefined {
    return this.allModels.find((m) => m.id === modelId);
  }

  loadActiveCache(): Record<string, unknown> {
    try {
      const data = fs.readFileSync(this.activeCacheFile, 'utf-8');
      return JSON.parse(data);
    } catch {
      return {};
    }
  }

  loadAllCache(): Record<string, unknown> {
    try {
      const data = fs.readFileSync(this.cacheFile, 'utf-8');
      return JSON.parse(data);
    } catch {
      return {};
    }
  }

  summary(): ModelSummary {
    return {
      total_models: this.allModels.length,
      free_models: this.freeModels.length,
      active_free_models: this.activeFreeModels.length,
      last_refresh: this.lastRefresh ? new Date(this.lastRefresh).toISOString() : 'never',
      last_test: this.lastTest ? new Date(this.lastTest).toISOString() : 'never',
    };
  }
}

// CLI entry point
const isMainModule = (() => {
  if (typeof process !== 'undefined' && process.argv) {
    const scriptPath = process.argv[1] || '';
    return scriptPath.endsWith('openrouter_free_models.ts') || scriptPath.endsWith('openrouter_free_models.js');
  }
  return false;
})();

if (isMainModule) {
  const args = process.argv.slice(2);
  const command = args[0];

  const getApiKey = (): string | undefined => {
    const apiKeyIdx = args.indexOf('--api-key');
    if (apiKeyIdx !== -1 && args[apiKeyIdx + 1]) {
      return args[apiKeyIdx + 1];
    }
    return process.env.OPENROUTER_API_KEY;
  };

  const run = async () => {
    const apiKey = getApiKey();
    if (!apiKey) {
      console.error('Error: OPENROUTER_API_KEY not set. Provide via --api-key or environment variable.');
      process.exit(1);
    }

    const manager = new OpenRouterFreeModels(apiKey);

    switch (command) {
      case 'refresh': {
        console.log('Fetching all models from OpenRouter...');
        const free = await manager.refreshAllModels();
        console.log(`Found ${free.length} free models out of ${manager.allModels.length} total`);
        break;
      }

      case 'list': {
        const cached = manager.loadAllCache();
        if (cached.free_models_count && (cached as any).free_models) {
          const freeModels = (cached as any).free_models as ModelInfo[];
          console.log(`Free models (${freeModels.length}):`);
          for (const m of freeModels) {
            console.log(`  ${m.id} - ${m.name} (context: ${m.context_length})`);
          }
        } else {
          console.log('No models cached. Run "refresh" first.');
        }
        break;
      }

      case 'test': {
        const concurrentIdx = args.indexOf('--concurrent');
        const maxConcurrent = concurrentIdx !== -1 && args[concurrentIdx + 1]
          ? parseInt(args[concurrentIdx + 1], 10)
          : undefined;

        console.log('Testing all free models...');
        const results = await manager.testFreeModels({ maxConcurrent });
        console.log(
          `\nResults: ${results.active_count} active / ${results.failed_count} failed / ${results.total_tested} tested`
        );
        if (results.active.length > 0) {
          console.log('\nActive free models:');
          for (const m of results.active) {
            console.log(`  ${m.model_id} (${m.response_time_ms}ms)`);
          }
        }
        break;
      }

      case 'test-single': {
        const modelId = args[1];
        if (!modelId) {
          console.error('Usage: node openrouter_free_models.ts test-single <model_id>');
          process.exit(1);
        }
        console.log(`Testing model: ${modelId}`);
        const results = await manager.testFreeModels({ modelIds: [modelId] });
        if (results.active.length > 0) {
          const m = results.active[0];
          console.log(`  ACTIVE: ${m.model_id} (${m.response_time_ms}ms)`);
          console.log(`  Response: ${m.test_response.substring(0, 100)}`);
        } else {
          console.log(`  FAILED: ${modelId}`);
        }
        break;
      }

      case 'active': {
        try {
          const data = fs.readFileSync(manager.activeCacheFile, 'utf-8');
          const models = JSON.parse(data) as ActiveFreeModel[];
          console.log(`Active free models (${models.length}):`);
          for (const m of models) {
            console.log(`  ${m.model_id} - ${m.name} (${m.response_time_ms}ms)`);
          }
        } catch {
          console.log('No active models cached. Run "test" first.');
        }
        break;
      }

      case 'summary': {
        const s = manager.summary();
        console.log(JSON.stringify(s, null, 2));
        break;
      }

      default:
        console.log(`Usage: node openrouter_free_models.ts {refresh|list|test|test-single <model_id>|active|summary} [--api-key <key>] [--concurrent <n>]`);
        process.exit(1);
    }
  };

  run().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: dotfiles-ai/skills/openrouter_free_models/openrouter_free_models.py
#!/usr/bin/env python3
"""
OpenRouter Free Models Manager

Fetches all models from OpenRouter, filters for free models,
and tests each one to maintain an active list of working free models.

Usage:
    from openrouter_free_models import OpenRouterFreeModels

    manager = OpenRouterFreeModels(api_key="sk-or-...")
    await manager.refresh_all_models()
    await manager.test_free_models()
    active = manager.get_active_free_models()
"""

import asyncio
import json
import os
import time
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Optional


@dataclass
class ModelInfo:
    id: str
    name: str
    context_length: int
    pricing_prompt: float
    pricing_completion: float
    architecture: dict
    top_provider: dict
    supported_parameters: list
    created: int
    description: str = ""
    hugging_face_id: Optional[str] = None


@dataclass
class ActiveFreeModel:
    model_id: str
    name: str
    context_length: int
    tested_at: float
    response_time_ms: float
    test_prompt: str = ""
    test_response: str = ""
    status: str = "active"


class OpenRouterFreeModels:
    BASE_URL = "https://openrouter.ai/api/v1"
    MODELS_ENDPOINT = f"{BASE_URL}/models"
    CHAT_ENDPOINT = f"{BASE_URL}/chat/completions"

    TEST_PROMPT = "Reply with exactly: PING"
    TEST_TIMEOUT = 30
    CONCURRENT_TESTS = 5

    def __init__(
        self,
        api_key: Optional[str] = None,
        cache_file: Optional[str] = None,
        active_cache_file: Optional[str] = None,
    ):
        self.api_key = api_key or os.environ.get("OPENROUTER_API_KEY", "")
        if not self.api_key:
            raise ValueError(
                "OPENROUTER_API_KEY not set. Provide via constructor or environment variable."
            )

        cache_dir = Path.home() / ".cache" / "openrouter"
        cache_dir.mkdir(parents=True, exist_ok=True)

        self.cache_file = cache_file or str(cache_dir / "all_models.json")
        self.active_cache_file = active_cache_file or str(cache_dir / "active_free_models.json")

        self.all_models: list[ModelInfo] = []
        self.free_models: list[ModelInfo] = []
        self.active_free_models: list[ActiveFreeModel] = []
        self._last_refresh: float = 0
        self._last_test: float = 0

    def _headers(self) -> dict:
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "HTTP-Referer": "https://github.com/cbwinslow/dotfiles",
            "X-Title": "CBW OpenRouter Free Models",
        }

    async def _fetch_json(self, session, url: str, method: str = "GET", body: Optional[dict] = None) -> dict:
        kwargs = {
            "method": method,
            "url": url,
            "headers": self._headers(),
        }
        if body:
            kwargs["json"] = body

        async with session.request(**kwargs) as resp:
            if resp.status != 200:
                return None
            return await resp.json()

    async def refresh_all_models(self) -> list[ModelInfo]:
        """Fetch all models from OpenRouter API and cache them."""
        import aiohttp

        timeout = aiohttp.ClientTimeout(total=60)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(self.MODELS_ENDPOINT, headers=self._headers()) as resp:
                resp.raise_for_status()
                data = await resp.json()

        models_data = data.get("data", [])
        self.all_models = []

        for m in models_data:
            pricing = m.get("pricing", {})
            arch = m.get("architecture", {})
            top_prov = m.get("top_provider", {})

            model_info = ModelInfo(
                id=m["id"],
                name=m.get("name", m["id"]),
                context_length=m.get("context_length", 0),
                pricing_prompt=float(pricing.get("prompt", 0)),
                pricing_completion=float(pricing.get("completion", 0)),
                architecture=arch,
                top_provider=top_prov,
                supported_parameters=m.get("supported_parameters", []),
                created=m.get("created", 0),
                description=m.get("description", ""),
                hugging_face_id=m.get("hugging_face_id"),
            )
            self.all_models.append(model_info)

        self._filter_free_models()
        self._save_cache()
        self._last_refresh = time.time()

        return self.free_models

    def _filter_free_models(self):
        """Filter models where both prompt and completion pricing are 0."""
        self.free_models = [
            m for m in self.all_models
            if m.pricing_prompt == 0 and m.pricing_completion == 0
        ]

    def _save_cache(self):
        """Cache all models and free models to disk."""
        cache = {
            "refreshed_at": self._last_refresh,
            "total_models": len(self.all_models),
            "free_models_count": len(self.free_models),
            "free_models": [asdict(m) for m in self.free_models],
        }
        with open(self.cache_file, "w") as f:
            json.dump(cache, f, indent=2)

    def _save_active_cache(self):
        """Save active free models to disk."""
        cache = {
            "tested_at": self._last_test,
            "active_count": len(self.active_free_models),
            "failed_count": len(getattr(self, "_failed_models", [])),
            "active_models": [asdict(m) for m in self.active_free_models],
            "failed_models": getattr(self, "_failed_models", []),
        }
        with open(self.active_cache_file, "w") as f:
            json.dump(cache, f, indent=2)

    async def _test_single_model(self, session, model_id: str, name: str, context_length: int) -> Optional[ActiveFreeModel]:
        """Test a single free model with a simple chat completion."""
        start = time.time()
        try:
            async with session.post(
                self.CHAT_ENDPOINT,
                headers=self._headers(),
                json={
                    "model": model_id,
                    "messages": [{"role": "user", "content": self.TEST_PROMPT}],
                    "max_tokens": 10,
                },
            ) as resp:
                elapsed_ms = (time.time() - start) * 1000

                if resp.status != 200:
                    return None

                text = await resp.text()
                text = text.strip()
                if not text:
                    return None

                data = json.loads(text)
                choice = data.get("choices", [{}])[0]
                message = choice.get("message", {})
                content = message.get("content", "") or message.get("reasoning", "")

                if not content:
                    return None

                return ActiveFreeModel(
                    model_id=model_id,
                    name=name,
                    context_length=context_length,
                    tested_at=time.time(),
                    response_time_ms=round(elapsed_ms, 2),
                    test_prompt=self.TEST_PROMPT,
                    test_response=content[:200],
                    status="active",
                )
        except Exception:
            return None

    async def test_free_models(
        self,
        model_ids: Optional[list[str]] = None,
        max_concurrent: int = CONCURRENT_TESTS,
    ) -> dict:
        """
        Test all free models (or a subset) to see which are working.

        Args:
            model_ids: Specific model IDs to test. If None, tests all free models.
            max_concurrent: Max concurrent test requests.

        Returns:
            dict with 'active' and 'failed' model lists.
        """
        if not self.free_models:
            await self.refresh_all_models()

        targets = [m for m in self.free_models if not model_ids or m.id in model_ids]

        self.active_free_models = []
        self._failed_models = []

        import aiohttp
        timeout = aiohttp.ClientTimeout(total=self.TEST_TIMEOUT)
        connector = aiohttp.TCPConnector(limit=0, force_close=True)

        async with aiohttp.ClientSession(timeout=timeout, connector=connector) as session:
            semaphore = asyncio.Semaphore(max_concurrent)

            async def bounded_test(model: ModelInfo):
                async with semaphore:
                    return await self._test_single_model(session, model.id, model.name, model.context_length)

            tasks = [bounded_test(m) for m in targets]
            results = await asyncio.gather(*tasks, return_exceptions=True)

            for result in results:
                if isinstance(result, Exception) or result is None:
                    continue
                if isinstance(result, ActiveFreeModel):
                    self.active_free_models.append(result)

            for m in targets:
                if not any(a.model_id == m.id for a in self.active_free_models):
                    self._failed_models.append({
                        "model_id": m.id,
                        "name": m.name,
                        "context_length": m.context_length,
                    })

        self._last_test = time.time()
        self._save_active_cache()

        return {
            "active": [asdict(m) for m in self.active_free_models],
            "failed": self._failed_models,
            "total_tested": len(targets),
            "active_count": len(self.active_free_models),
            "failed_count": len(self._failed_models),
        }

    def get_active_free_models(self) -> list[ActiveFreeModel]:
        """Return the current list of active free models."""
        return self.active_free_models

    def get_free_model_ids(self) -> list[str]:
        """Return IDs of all free models."""
        return [m.id for m in self.free_models]

    def get_model_by_id(self, model_id: str) -> Optional[ModelInfo]:
        """Look up a model by its ID."""
        for m in self.all_models:
            if m.id == model_id:
                return m
        return None

    def load_active_cache(self) -> dict:
        """Load previously cached active free models."""
        try:
            with open(self.active_cache_file) as f:
                return json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            return {}

    def load_all_cache(self) -> dict:
        """Load previously cached all models."""
        try:
            with open(self.cache_file) as f:
                return json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            return {}

    def summary(self) -> dict:
        """Print a summary of the current state."""
        return {
            "total_models": len(self.all_models),
            "free_models": len(self.free_models),
            "active_free_models": len(self.active_free_models),
            "last_refresh": time.ctime(self._last_refresh) if self._last_refresh else "never",
            "last_test": time.ctime(self._last_test) if self._last_test else "never",
        }


async def main():
    """CLI entry point."""
    import argparse

    parser = argparse.ArgumentParser(description="OpenRouter Free Models Manager")
    parser.add_argument("--api-key", help="OpenRouter API key")
    parser.add_argument("--refresh", action="store_true", help="Refresh all models from API")
    parser.add_argument("--test", action="store_true", help="Test all free models")
    parser.add_argument("--model", help="Test a specific model ID")
    parser.add_argument("--list", action="store_true", help="List free models")
    parser.add_argument("--active", action="store_true", help="Show active free models")
    parser.add_argument("--summary", action="store_true", help="Show summary")
    parser.add_argument("--concurrent", type=int, default=5, help="Max concurrent tests")
    args = parser.parse_args()

    try:
        manager = OpenRouterFreeModels(api_key=args.api_key)
    except ValueError as e:
        print(f"Error: {e}")
        return

    if args.refresh:
        print("Fetching all models from OpenRouter...")
        free = await manager.refresh_all_models()
        print(f"Found {len(free)} free models out of {len(manager.all_models)} total")

    if args.test:
        print("Testing all free models...")
        results = await manager.test_free_models(max_concurrent=args.concurrent)
        print(f"\nResults: {results['active_count']} active / {results['failed_count']} failed / {results['total_tested']} tested")
        if results["active"]:
            print("\nActive free models:")
            for m in results["active"]:
                print(f"  {m['model_id']} ({m['response_time_ms']}ms)")

    if args.model:
        print(f"Testing model: {args.model}")
        results = await manager.test_free_models(model_ids=[args.model])
        if results["active"]:
            m = results["active"][0]
            print(f"  ACTIVE: {m['model_id']} ({m['response_time_ms']}ms)")
            print(f"  Response: {m['test_response'][:100]}")
        else:
            print(f"  FAILED: {args.model}")

    if args.list:
        if not manager.free_models:
            cached = manager.load_all_cache()
            if cached.get("free_models"):
                print(f"Loaded {len(cached['free_models'])} free models from cache")
                for m in cached["free_models"]:
                    print(f"  {m['id']} - {m['name']} (ctx: {m['context_length']})")
            else:
                print("No models cached. Run --refresh first.")
        else:
            print(f"Free models ({len(manager.free_models)}):")
            for m in manager.free_models:
                print(f"  {m.id} - {m.name} (ctx: {m.context_length})")

    if args.active:
        cached = manager.load_active_cache()
        if cached.get("active_models"):
            print(f"Active free models ({cached['active_count']}):")
            for m in cached["active_models"]:
                print(f"  {m['model_id']} - {m['name']} ({m['response_time_ms']}ms)")
        else:
            print("No active models cached. Run --test first.")

    if args.summary:
        s = manager.summary()
        print(json.dumps(s, indent=2))


if __name__ == "__main__":
    asyncio.run(main())
